### DISCUSSION TITLE
How to route when use `{Literals}` at last?

### DISCUSSION THREAD

**Discussion by adelvadia** [ThreeMammals/Ocelot#2095] [Q&A] [answered]
I have multiple route configurations in ocelot.json

```
{
     "UpstreamPathTemplate": "/{Version}/List/{ListId}",
     "DownstreamPathTemplate": "/Api/{Version}/List/{ListId}",
     ...
     ...
},
{
     "UpstreamPathTemplate": "/{Version}/List/{ListId}/View/{ViewId}/Records",
     "DownstreamPathTemplate": "/Api/{Version}/List/{ListId}/View/{ViewId}/Records",
     ...
     ...
}
```

GET http://localhost/V1/List/100/View/256/Records

Above GET request matches first route instead of second one.

I can resolve this issue by specifying high priority to second one. Here, I expect best match instead of first match. Is there any solution to do so?

## Specifications

  - Version: 23.2.2
  - Platform: Winodws 0x64
  - Subsystem: N.A

**Comment by raman-m** [comment, marked as answer]
> I can resolve this issue by specifying high priority to second one. Here, I expect best match instead of first match.
**Is there any solution to do so?**

Yes, There is!

----
> Above GET request matches first route instead of second one.

Indeed, the GET request matches the first route rather than the second one because the first route is a Catch All route. A Catch All route is a unique type of routing that attempts to match all content in the last placeholder, including query strings.
I believe that when multiple routes match a URL, the first one is selected.

----
> I can resolve this issue by specifying high priority to second one.

Certainly, this is the correct solution. Prioritizing routing is necessary. However, I'm puzzled that implicit priorities were not applied to the CatchAll route, which should have the lowest priority. Perhaps we need to review our design, because the `/{everything}` CatchAll must function; if not, then we have a bug.

----
> Here, I expect best match instead of first match. Is there any solution to do so?

Seems your route template can be shorter:
```json
{
     "UpstreamPathTemplate": "/{everything}",
     "DownstreamPathTemplate": "/Api/{everything}",
     ...
}
```
I don't understand, why did you define so long template. Possibly you need to read [Routing](https://ocelot.readthedocs.io/en/latest/features/routing.html) docs once again.

### Recommendation
Define one  [Catch All](https://ocelot.readthedocs.io/en/latest/features/routing.html#catch-all) route! This should solve the problems.
```json
{
     "UpstreamPathTemplate": "/{everything}",
     "DownstreamPathTemplate": "/Api/{everything}",
     ...
}
```
Hope it helps!
